### ISSUE TITLE
Cross Platform Support

### ISSUE THREAD

**Issue by vpenades** [bepu/bepuphysics2#30]
Given that the whole code of Bepu2 is Net.Standard, theoretically it is cross platform by default, which is great!.... but...

I am seeing lots and lots of extreme optimizations, like using heavy duty vector math operations, and lots of memory tricks with Span<T> and even using unmanaged memory for storing buffers.

I've been contributing to SixLabots.ImageSharp, which are also doing lots of these performance optimizations... and they're already having a lot of issues on Xamarin/Android and UWP due to some optimizations not being available, or the compiler not optimizing as expected, or even producing invalid code.

Maybe you could consider, at some point in the roadmap, to introduce/port the demo to xamarin and/or UWP, just for the sake of leveraging and balancing Bepu2 for platforms other than desktop.

UWP port is probably quite trivial, since UWP is supported by SharpDX; the interesting thing is that UWP uses the Roslyn AOT compiler which gives a quite impressive performance boost, but also is very buggy with Span<T> and System.Numerics.Vectors.

Android/IOS would probably need an OpenGL rendering head, which might probe more complicated.

**Comment by RossNordby** [comment]
It's true that non-RyuJIT compilers are going to have performance issues with v2, at least to begin with. I haven't done thorough benchmarks on other runtimes, but I did test an older version of mono and it's about as bad as you'd expect.

It's also likely that some other platforms will indeed have correctness issues. I've had to dodge a few of those even with CoreCLR due to some of the questionable things I've done for performance.

But I probably won't port the demos themselves to multiple platforms in the foreseeable future. There isn't much value in it- correctness can be verified without bringing over the demos infrastructure.

I will do a correctness pass just to make sure the engine itself actually runs on the most common platforms at some point. I'm hesitant to spend time actually creating platform specific optimizations, though. CoreCLR (and CoreRT when it becomes mature enough) are the most interesting platforms for my own purposes, and the future of other platforms is in flux. In particular, the elephant in the room, Unity, is adding a whole new compiler target ([plus some other question marks](http://xoofx.com/blog/2018/04/06/porting-unity-to-coreclr/)).

In other words, I'd like to wait and see before committing to anything significant. I'll probably update the roadmap with regard to platform support once things are a little more concrete.

**Comment by RossNordby** [comment]
I didn't do particularly thorough testing, but it appears to at least work on some other platforms, so I'm going to assume that everything is 100% totally fine until proven otherwise.
